components/core/src/clp_s/Schema.hpp

```diff
@@ -0,0 +1,71 @@
+#ifndef CLP_S_SCHEMA_HPP
+#define CLP_S_SCHEMA_HPP
+
+#include <cstddef>
+#include <cstdint>
+#include <vector>
+
+namespace clp_s {
+/**
+ * Class representing a a schema made up of MST
+ */
+class Schema {
+public:
+    using schema_t = std::vector<int32_t>;
+
+    /**
+     * Insert a single node into the ordered region of the schema.
+     */
+    void insert_ordered(int32_t mst_node_id);
+
+    /**
+     * Insert a single node into the unordered region of the schema.
+     */
+    void insert_unordered(int32_t mst_node_id);
+
+    /**
+     * Insert an unordered list of nodes into the unordered region of the schema.
+     */
+    void insert_unordered(schema_t const& mst_node_ids);
+
+    /**
+     * Clear the Schema object so that it can be reused without reallocating the underlying vector.
+     */
+    void clear() {
+        m_schema.clear();
+        m_num_unordered = 0;
+    }
+
+    /**
+     * @return the number of elements in the underlying schema
+     */
+    size_t size() const { return m_schema.size(); }
+
+    /**
+     * @return iterators to the underlying schema
+     */
+    auto begin() { return m_schema.begin(); }
+
+    auto end() { return m_schema.end(); }
+
+    /**
+     * @return constant iterators to the underlying schema
+     */
+    auto begin() const { return m_schema.cbegin(); }
+
+    auto end() const { return m_schema.cend(); }
```

Comment:
You would think, but surprisingly that doesn't seem to work with the for (auto item: schema) syntax.

Reply:
Oh I see, it's because the ranged-based for seems to look for a method named `begin` and doesn't consider `cbegin`. The way `std::vector` solves this is by defining a non-const `begin`, const `begin`, and const `cbegin`. Can we do the same here?

Reply:
Sure, I'll add cbegin/cend as well